pkg/alerter/alerter.go

```diff
@@ -0,0 +1,227 @@
+			}
+		}()
+	}
+
+	// Subroutine to concatenate errors together
+	resultCh := make(chan error)
+	go func() {
+		defer close(resultCh)
+		var errs []string
+		for err := range errCh {
+			if err == nil {
+				continue
+			}
+			errs = append(errs, err.Error())
+		}
+		if n := len(errs); n > 0 {
+			resultCh <- fmt.Errorf("failed to update %d dashboards: %v", n, strings.Join(errs, ", "))
+		}
+		resultCh <- nil
+	}()
+
+	// Queue up all the dashboards
+	for _, d := range cfg.Dashboards {
+		if dashboard != "" && dashboard != d.Name {
+			logrus.WithField("dashboard", d.Name).Info("Skipping")
+			continue
+		}
+		dashboards <- d
+	}
+	close(dashboards)
+	wg.Wait() //wait for them to stop reporting
+	close(errCh)
+	return <-resultCh
+}
+
+// Paths to relevant GCS files
+var (
+	normalizer = regexp.MustCompile(`[^a-z0-9]+`)
+)
+
+func alerterPath(name string) string {
+	return "alerter-" + normalizer.ReplaceAllString(strings.ToLower(name), "")
+}
+
+func writeSummary(ctx context.Context, client *storage.Client, path gcs.Path, sum *summarypb.DashboardSummary) error {
+	buf, err := proto.Marshal(sum)
+	if err != nil {
+		return fmt.Errorf("marshal: %v", err)
+	}
+	return gcs.Upload(ctx, client, path, buf, gcs.DefaultAcl, "no-cache")
+}
+
+func readSummary(ctx context.Context, client *storage.Client, path gcs.Path) (*summarypb.DashboardSummary, error) {
+	r, err := client.Bucket(path.Bucket()).Object(path.Object()).NewReader(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open config: %v", err)
+	}
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config: %v", err)
+	}
+	var summary summarypb.DashboardSummary
+	if err = proto.Unmarshal(buf, &summary); err != nil {
+		return nil, fmt.Errorf("failed to parse: %v", err)
+	}
+	return &summary, nil
+}
+
+func callOutDifferences(ctx context.Context, new, old *summarypb.DashboardSummary, config *configpb.Dashboard, ConsistentFailure, Stale Klaxon) {
+	// Deserialize dashboard summaries into maps for easy access
+	oldSummaries := map[string]*summarypb.DashboardTabSummary{}
+	newSummaries := map[string]*summarypb.DashboardTabSummary{}
+
+	for _, tabSummary := range old.TabSummaries {
+		oldSummaries[tabSummary.DashboardTabName] = tabSummary
+	}
+
+	for _, tabSummary := range new.TabSummaries {
+		newSummaries[tabSummary.DashboardTabName] = tabSummary
+	}
+
+	for _, dashboardTab := range config.DashboardTab {
+		newSummary := newSummaries[dashboardTab.Name]
+		oldSummary := oldSummaries[dashboardTab.Name]
+
+		if newSummary.OverallStatus == summarypb.DashboardTabSummary_STALE && oldSummary.OverallStatus != summarypb.DashboardTabSummary_STALE {
+			err := Stale(ctx, newSummary, oldSummary, dashboardTab)
+			if err != nil {
+				logrus.Errorf("OnBecomingStale Dashboard %s Error: %v", dashboardTab.Name, err)
+			}
+		}
+
+		alertThreshold := dashboardTab.AlertOptions.NumFailuresToAlert
+
+		for _, failingTest := range newSummary.FailingTestSummaries {
```

Comment:
I'm not sure what this logic is handling. If the test is at the alert threshold in the old AND new summary, don't alert?